pkg/controller/workload/kubernetes/resource/resource.go

```diff
@@ -244,31 +257,39 @@ type unstructuredClient struct {
 	kube client.Client
 }
 
-func (c *unstructuredClient) sync(ctx context.Context, template *unstructured.Unstructured) (*v1alpha1.RemoteStatus, error) {
+func (c *unstructuredClient) sync(ctx context.Context, data []byte) (*v1alpha1.RemoteStatus, error) {
 	// We make another copy of our template here so we can compare the template
 	// as passed to this method with the remote resource.
+	template := &unstructured.Unstructured{}
+	if err := json.Unmarshal(data, template); err != nil {
+		return nil, errors.Wrap(err, errUnmarshalTemplate)
+	}
+
 	remote := template.DeepCopy()
+
 	key, err := client.ObjectKeyFromObject(remote)
 	if err != nil {
-		return nil, errors.Wrap(err, "cannot take object key from resource")
+		return nil, errors.Wrap(err, errGetKey)
 	}
+
+	// TODO(hasheddan): this Get operation and subsequent controller check can
+	// be eliminated once resource.Apply is refactored to accept arbitrary
+	// ApplyOption functions.
```

Comment:
Note for anyone else reading this: it wasn't immediately obvious to me why resource.Apply doesn't accept arbitrary ApplyOption functions today, until I recalled that ApplyOptions are applied to an options struct, which controls flow within the function.